js/src/modals/Shapeshift/shapeshift.js

```diff
@@ -0,0 +1,320 @@
+  setFatalError (message) {
+    this.setState({
+      stage: 0,
+      error: {
+        fatal: true,
+        message
+      }
+    });
+  }
+
+  onClose = () => {
+    this.setStage(0);
+    this.props.onClose && this.props.onClose();
+  }
+
+  onShift = () => {
+    const { address, newError } = this.props;
+    const { coinPair, refundAddress } = this.state;
+
+    this.setState({
+      stage: 1,
+      shifting: true
+    });
+
+    shapeshift
+      .shift(address, refundAddress, coinPair)
+      .then((result) => {
+        console.log('onShift', result);
+        const depositAddress = result.deposit;
+
+        shapeshift.subscribe(depositAddress, this.onExchangeInfo);
+        this.setState({ depositAddress });
+      })
+      .catch((error) => {
+        console.error('onShift', error);
+        const message = `Failed to start exchange: ${error.message}`;
+
+        newError(new Error(message));
+        this.setFatalError(message);
+      });
+  }
+
+  onChangeSymbol = (event, coinSymbol) => {
+    const coinPair = `${coinSymbol.toLowerCase()}_eth`;
+
+    this.setState({
+      coinPair,
+      coinSymbol,
+      price: null
+    });
+    this.getPrice(coinPair);
+  }
+
+  onChangeRefund = (event, refundAddress) => {
+    this.setState({ refundAddress });
+  }
+
+  onToggleAccept = () => {
+    const { hasAccepted } = this.state;
+
+    this.setState({
+      hasAccepted: !hasAccepted
+    });
+  }
+
+  onExchangeInfo = (error, result) => {
+    const { newError } = this.props;
+
+    if (error) {
+      console.error('onExchangeInfo', error);
+
+      if (error.fatal) {
+        this.setFatalError(error.message);
+      }
+
+      newError(error);
+      return;
+    }
+
+    console.log('onExchangeInfo', result.status, result);
+
+    switch (result.status) {
+      case 'received':
+        this.setState({ depositInfo: result });
+        this.setStage(2);
+        return;
+
+      case 'complete':
+        this.setState({ exchangeInfo: result });
+        this.setStage(3);
+        return;
+    }
+  }
+
+  getPrice (coinPair) {
+    shapeshift
+      .getMarketInfo(coinPair)
+      .then((price) => {
+        this.setState({ price });
+      })
+      .catch((error) => {
+        console.error('getPrice', error);
+      });
+  }
+
+  retrieveCoins () {
+    const { newError } = this.props;
+    const { coinPair } = this.state;
+
+    shapeshift
+      .getCoins()
+      .then((_coins) => {
+        const coins = Object.values(_coins).filter((coin) => coin.status === 'available');
+
+        this.getPrice(coinPair);
+        this.setState({ coins });
+      })
+      .catch((error) => {
+        console.error('retrieveCoins', error);
+        const message = `Failed to retrieve available coins from ShapeShift.io: ${error.message}`;
+
+        newError(new Error(message));
+        this.setFatalError(message);
+      });
+  }
+}
+
+function mapStateToProps (state) {
+  return {};
+}
+
+function mapDispatchToProps (dispatch) {
```

Comment:
If this was a component like a dapp, self-contained, yes. I honestly don't want to be coding for the next 2 days just to do that complexity and conversion where the state is only applicable to the single component.

So no, I'd rather get features out than coding for the sake of coding where in the end I won't be able to follow the actual code. I think Redux is great where I need global state, don't need it here. (newError is global)